service/history/api/multioperation/api.go

```diff
@@ -108,64 +110,120 @@ func Invoke(
 		// TODO: send per-operation error details
 		return nil, fmt.Errorf("failed create workflow starter: %w", err)
 	}
+	workflowCache := workflowConsistencyChecker.GetWorkflowCache()
+	return startAndUpdateWorkflow(ctx, shardContext, workflowCache, starter, updater)
+}
 
-	var workflowCtx workflow.Context
-	var updateErr error
-	updater := updateworkflow.NewUpdater(
+func updateWorkflow(
+	ctx context.Context,
+	shardContext shard.Context,
+	currentWorkflowLease api.WorkflowLease,
+	updater *updateworkflow.Updater,
+) (*historyservice.ExecuteMultiOperationResponse, error) {
+	startOpResp := &historyservice.ExecuteMultiOperationResponse_Response{
+		Response: &historyservice.ExecuteMultiOperationResponse_Response_StartWorkflow{
+			StartWorkflow: &historyservice.StartWorkflowExecutionResponse{
+				RunId:   currentWorkflowLease.GetContext().GetWorkflowKey().RunID,
+				Started: false,
+			},
+		},
+	}
+
+	// apply update to workflow
+	err := api.UpdateWorkflowWithNew(
 		shardContext,
-		workflowConsistencyChecker,
-		matchingClient,
-		updateReq,
-	)
-	startResp, err := starter.Invoke(
 		ctx,
-		func(lease api.WorkflowLease) error {
-			workflowCtx = lease.GetContext()
-			_, updateErr = updater.ApplyRequest(ctx, lease)
-			return updateErr
+		currentWorkflowLease,
+		func(lease api.WorkflowLease) (*api.UpdateWorkflowAction, error) {
+			return updater.ApplyRequest(ctx, lease)
 		},
+		nil,
 	)
+
+	// release lock!
+	currentWorkflowLease.GetReleaseFn()(err)
+
 	if err != nil {
-		// TODO: send per-operation error details
-		return nil, fmt.Errorf("failed to start workflow: %w", err)
+		return onUpdateError(err, updater, startOpResp)
+	}
+
+	// wait for the update to complete
+	updateResp, err := updater.OnSuccess(ctx)
+	if err != nil {
+		// TODO: send Update outcome failure instead
+		return nil, fmt.Errorf("failed to complete Workflow Update: %w", err)
```

Comment:
I see `TODO` here, and I think `serviceerror.NewUnavailable()` is better for temporary solution.